manager/state/raft/raft.go

```diff
@@ -1302,6 +1303,43 @@ func (n *Node) reportNewAddress(ctx context.Context, id uint64) error {
 	return n.transport.UpdatePeerAddr(id, newAddr)
 }
 
+// RaftMessageStream is the server endpoint for streaming Raft messages.
+func (n *Node) RaftMessageStream(stream api.Raft_RaftMessageStreamServer) error {
+	var recvdMsg, assembledMessage *api.ProcessRaftMessageRequest
+	var err error
+	for {
+		recvdMsg, err = stream.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			log.G(context.Background()).WithError(err).Error("error while reading from stream")
+			return err
+		}
+
+		if assembledMessage == nil {
+			assembledMessage = recvdMsg
+			continue
+		}
+
+		// Append the received snapshot data.
+		if recvdMsg.Message.Type == raftpb.MsgSnap {
+			assembledMessage.Message.Snapshot.Data = append(assembledMessage.Message.Snapshot.Data, recvdMsg.Message.Snapshot.Data...)
```

Comment:
Doesn't this arrive at the same problem? This buffers the whole thing into a single buffer. How does this address the issue?

Reply:
The problem was being able to stream large snapshots without arbitrarily large gRPC message limits and timeout outs. This addresses that problem. Memory constraint is not the problem we're trying to solve, at least yet.

Reply:
@anshulpundir Fair enough. :)

This could double (or more!) memory usage.

Here's how we can decide: Do we have to load the entire snapshot into memory to apply it? If yes, then this is fine, but if not, we should buffer this to disk (tmp is fine).

Reply:
Buffering to disk won't be any better would it ? to apply we'd still need to load the snapshot into memory before passing it to etcd/raft. 

Also, it seems MemoryStorage::ApplySnapshot() doesn't actually do a memory copy, so we should still just need 1x memory.

Reply:
There are three components:

1. "The bytes" of the snapshot itself.
2. The structure that gets applied to the store.
3. The store itself (not raft, but the database).

With the current model, I think the minimum is a 2x bump. Ideally, we can stream it in and avoid 1 and 2. Buffering to disk avoids 1, if 2 is necessary.

Looking at the code, there are two components: the entries and the snapshot data, a `[]byte`. From the looks of it, the snapshot is applied always as a `[]byte`, so we can never avoid 2. If a snapshot is applied in this fashion, there may exist a snapshot that has a similar size, which can contribute to bloat.

So, I think I agree with you: this code is probably fine as is, as long as this doesn't get copied (and it doesn't look like it does). Following that, I don't think there is a easy way to avoid the 2x memory bump by applying a snapshot in a stream.

Reply:
i might be mistaken, but I don't think this code is any worse off than the regular raft message rpc call, except for the allocations and copies required to expand the array from the successive `append`s.

Reply:
that said, maybe it's advantageous to take the couple of extra bytes to send the total message length with each stream request, so we can on the first message allocate an appropriately sized byte slice.

Reply:
I did consider that, but streaming snapshots should not happen that frequently, so this is probably OK for now. We can always make the change to send the size later.